### PR TITLE
#2649 Google OAuth OOB migration (Using Content Script)

### LINKED ISSUE #2649: stop using urn:ietf:wg:oauth:2.0:oob:auto before Oct 3
> Any reason why the OAuth to Google uses "Programmatic extraction urn:ietf:wg:oauth:2.0:oob:auto" ? The docs from Google says this option is deprecated and should only be used for embedded browsers/web views

> Google OAuth does provide a way to pass OAuth token via postMessage or other safer means

https://cure53.slack.com/archives/GSMB8RGFK/p1583983921015100

### PR DESCRIPTION
**Pull request by ioanmo226**
This PR implemented google oauth oob migration.

close #2649 // if this PR closes an issue

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

### CODE REVIEW

**Comment by ioanmo226** [comment]
All finished. Now ready for review.
@tomholub Please check GitGuardian

**Comment by tomholub** [comment]
@ioanmo226 could you please manually test this change in Firefox - does it still work?

File: extension/js/common/oauth2/oauth2_inject.ts
Comment (tomholub) [inline extension/js/common/oauth2/oauth2_inject.ts:7]: Please update this comment to explain what type of privileged access this is, and what was it needed for that can't be done directly from here.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:197]: Method `getOAuthResult` should be called `renderOAuthWindowAndGetResult` to reflect what it actually does.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:174]: This `if (chrome.runtime.lastError) {` is rather indirect, but I suppose you had some reason for it? As opposed to just throwing the error where it happens and then catching it normally? (or in async lingo, rejecting the promise where it happens, and then catching that rejection normally in async try/catch.)

When this error happens, exactly where does it happen, which code generates it? It would be better to catch and handle the error right there.

File: test/source/mock/lib/api.ts
Comment (tomholub) [inline test/source/mock/lib/api.ts:238]: We can probably come up with a cleaner solution here. I suppose, in mocked API, we shouldn't need to deal with the `robots.txt` location at all. We already have a different destination for oauth2 dialog in mock, and we can also have a different place instead of google/robots.txt for the redirect, something that maybe wouldn't need this code here.

However, I haven't fully examined how this is used, therefore I'm not 100% sure if my suggestion is appropriate.

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (tomholub) [inline extension/js/common/oauth2/oauth2_finish.ts:4]: I suppose this is where the `lastError` may happen, as showing in

```ts
  public static finishAuth = (url: string) => {
    const views = chrome.extension.getViews();
    // Loop through existing extension views and excute any stored callbacks.
    for (const view of views) {
      if ((view as any)['oauth-callback']) {
        (view as any)['oauth-callback'](url); // tslint:disable-line:no-unsafe-any
        (view as any)['oauth-login-finished'] = true;
      }
    }
    chrome.runtime.lastError = undefined;
    window.open('', '_self', '');
    window.close();
  };
```

It's a rather creative solution :-) but you don't have to do this. To communicate between views, you can use browser messages (`BrowserMsg` class). In this case, you can send a broadcast message `BrowserMsg.send.authWindowResult('broadcast', { any JSON serializable object you need })`. You'll need to add this `authWindowResult` to `BrowserMsg`, and then also add appropriate listeners in the pages that need this, instead of setting 

```ts
    (window.top as any)['oauth-login-finished'] = false;
    (window.top as any)['oauth-callback'] = callback;
```

That will make it more consistent with the rest of the codebase.

Now that you did it, I wonder if there is some advantage to what you've done compared to what we're doing, but that would be a discussion for another time.

**Comment by ioanmo226** [comment]
It still works fine except oauth popup window doesn't close and show some javascript error in console.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:206]: Nice, so scopes can be obtained directly from the result.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:217]: In how many situations is this `save` used vs where is it not used? To try to have method with fewer disparate responsibilities, it may be worthwhile to remove it from here and explicitly save it where needed. But this depends how many places are we talking about. Since this is a private method, maybe it's just one place it's being called from? Then you could add the `if / save` there.

**Review by tomholub — CHANGES_REQUESTED** [review]
Looking good, and very happy we have a functioning solution.

Have you tried Firefox, and also Gmail live tests?

We just need to beat this code into submission through the PR review process in an attempt to simplify or increase clarity. I'll wait for @rrrooommmaaa review as well to also get his perspective.

**Comment by tomholub** [comment]
> It still works fine except oauth popup window doesn't close and show some javascript error in console.

Got it - that would also need to be addressed before merging.

**Comment by ioanmo226** [comment]
Gmail live tests are already broken in master branch and need to be fixed
https://github.com/FlowCrypt/flowcrypt-browser/issues/4594

> Have you tried Gmail live tests?

**Comment by tomholub** [comment]
> Gmail live tests are already broken in master branch and need to be fixed

I think we could do that in a separate PR after this PR merges. So you don't have to fix that here, since it was already broken prior.

File: extension/js/common/oauth2/oauth2_inject.ts
Comment (rrrooommmaaa) [inline extension/js/common/oauth2/oauth2_inject.ts:3]: typo. Should be `declared`

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (tomholub) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: @rrrooommmaaa wdyt?

File: test/source/mock/lib/api.ts
Comment (rrrooommmaaa) [reply test/source/mock/lib/api.ts:238]: Yes, looks like `https://www.google.com/robots.txt` is actually accessed (queried) in mock tests which we try to keep isolated.
This is not necessary as we have the code to replace URLs in mock manifests.
We're replacing `https://mail.google.com` with `https://gmail.localhost:8001` in https://github.com/FlowCrypt/flowcrypt-browser/blob/59bc0b3890294f05bd41f1bcf0eef6dbd25294ca/tooling/build-types-and-manifests.ts#L128
We can also replace `https://www.google.com` with `https://google.localhost:8001` (different host but still resolved to 127.0.0.1)

File: extension/js/common/oauth2/oauth2_inject.ts
Comment (rrrooommmaaa) [inline extension/js/common/oauth2/oauth2_inject.ts:6]: Have you tried to add `frame-ancestors https://www.google.com/robots.txt 'self'` (or the like) to manifest's `"content_security_policy"`. I guess it'll let you oauth2_finish right away.

We have this line `frame-ancestors https://mail.google.com 'self'` already and it is also overriden for mocks with https://github.com/FlowCrypt/flowcrypt-browser/blob/59bc0b3890294f05bd41f1bcf0eef6dbd25294ca/tooling/build-types-and-manifests.ts#L128

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: > If we use `BrowserMsg.send` in `oauth2.ts` and add listener in `GoogleAuth` we won't benefit callback(promise) feature.

Why, can't the message handler resolve a promise?

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: Let's talk about above code and let's say we use `BrowserMsg.send` to communicate between views. 
We might have `GoogleAuth.didGetOAuthResult` listener or something like that.
In `getOAuthResult` function, we launch oauth2 dialog by calling `GoogleAuth.oauthLogin` and we can't continue logic from here as we don't have success/failure callback.
Instead we have to add below logic into new listener method(GoogleAuth.didGetOAuthResult).

This way, google login logic will be seperate into `getOAuthResult` and `didGetOAuthResult` and it seems hard to read (at least for me)

```ts
if (!responseUrl) {
  return { acctEmail, result: 'Denied', error: 'Invalid response url', id_token: undefined };
}
const url = new URL(responseUrl!);
const allowedScopes = this.getParameterByName(url.search, 'scope');
const code = this.getParameterByName(url.search, 'code') ?? '';
if (!allowedScopes?.includes(this.OAUTH.scopes.compose) || !allowedScopes?.includes(this.OAUTH.scopes.modify)) {
...
```

File: extension/js/common/oauth2/oauth2_inject.ts
Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: Sorry but not sure exactly what you mean.
AFAIK, frame-ancestors directive specifies valid parents that may embed a page(using iframe or something like that).
I'm not sure how it would let oauth2_finish right away.

Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: What was the error preventing finishing OAuth2 directly from `robots.txt`?
There is a number of security options, maybe enabling any of them can help.

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: @ioanmo226 You're setting up a listener inside `launchWebAuthFlow`. Can you also set up a listener to a message that would be sent by `robots.txt`?

File: extension/js/common/oauth2/oauth2_inject.ts
Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: What I am trying to understand: what prevents us sending a message to close the oauth2 dialog directly from `robots.txt`?

Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: Like here:
```
// Even accepts arrays of patterns!
chrome.tabs.query({url: "*://*.example.com/*"}, function(tabs) {
  tabs.forEach(function(tab) {
    chrome.tabs.sendMessage(tab.id, message);
  });
});
```
https://stackoverflow.com/a/28832962

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: `robots.txt` page url contains redirect_url which contains `authorization_code`, `scopes`, etc.
Our final goal is to get this url when we call `GoogleAuth.oauthLogin`.
But without help of content script injection (`oauth2_inject.ts` for our case), we can't get that url.

Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: Yes, we get this data in `oauth2_inject.ts`. Can we send it to the tab that awaits it (has set up the listener for it?)

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: Can `robots.txt` page access `chrome.tabs` or `chrome.extension.getViews`?
I don't think it would.
> Like here:

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_inject.ts:6]: image

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: @tomholub , @rrrooommmaaa I tried to migrate to `BrowserMsg.send` and here's a commit. (Below code is temporary, like using localStorage to save datas)
https://github.com/FlowCrypt/flowcrypt-browser/pull/4665/commits/b682f9af7338628724bce4c4068fd200c9f4b290

To me, updated code looks super messy and hard to understand.
I prefer previous way using callback method and want to revert change.

Let me know what you think.

File: extension/js/common/oauth2/oauth2.ts
Comment (rrrooommmaaa) [inline extension/js/common/oauth2/oauth2.ts:27]: I meant something like this. Resolve the promise when the message is received. Wouldn't this work?

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2.ts:27]: Ohh.. Yeah.
It looks great. It's exactly what I wanted to achieve.
Thank you

File: extension/js/common/oauth2/oauth2_finish.ts
Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2_finish.ts:4]: Will mark this as resolved as @rrrooommmaaa pushed fix

File: extension/js/common/oauth2/oauth2.ts
Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2.ts:27]: This method has problem though which will only accept `auth_window_result` event once.
Like when user wants to change email in `Force Google Account email change`(in experiemental page), event doesn't get fired.
Finding solution now.

Comment (rrrooommmaaa) [reply extension/js/common/oauth2/oauth2.ts:27]: Yes, we need to remove both listeners when resolving the promise

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2.ts:27]: I thought experiemental `Force Google Account email change` feature doesn't work because event only fires once but that seems wrong.
It seems like event doesn't get fired in experimental page because it's wrapped in Iframe.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (ioanmo226) [reply extension/js/common/api/email-provider/gmail/google-auth.ts:210]: IMHO I think better to leave this as it is for now.
Only one place has `save` as false and as we also  save `scopes` in `googleAuthSaveTokens`, better to refactor this once we migrate to `load allowed scopes from google api` feature.

https://github.com/FlowCrypt/flowcrypt-browser/blob/37d466bf471bde88eae19a2ccd17f1bf74ce87d2/extension/js/common/api/email-provider/gmail/google-auth.ts#L295

image

File: extension/js/common/oauth2/oauth2.ts
Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2.ts:28]: @tomholub , @rrrooommmaaa Does anyone know why we use `BrowserMsg.send.bg.await.reconnectAcctAuthPopup` instead of using `GoogleAuth.newAuthPopup` directly?
Seems like UI test (enable google contact search test) fails because of this reason.

https://github.com/FlowCrypt/flowcrypt-browser/blob/37d466bf471bde88eae19a2ccd17f1bf74ce87d2/extension/chrome/elements/compose-modules/compose-recipients-module.ts#L725

Comment (tomholub) [reply extension/js/common/oauth2/oauth2.ts:28]: I don't see any good reason. Maybe some reason used to exist, but we refactored it away and it no longer makes sense. Feel free to remove `reconnectAcctAuthPopup`

Comment (ioanmo226) [reply extension/js/common/oauth2/oauth2.ts:28]: ok

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:123]: highlighting

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:255]: Has the CSRF check moved elsewhere? I see CSRF token getting created with the request but don't see the check anymore, it may be vulnerable to CSRF attacks.

**Review by tomholub — COMMENTED** [review]
This looks good to me. Will wait for Roman's approval / comments after the update.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [reply extension/js/common/api/email-provider/gmail/google-auth.ts:210]: Ok, please link this there to discuss further. Now I wonder, do we really need to do `save: false` in that situation? Maybe it wouldn't hurt to save there also?

**Comment by tomholub** [comment]
Is firefox behavior fixed now?

**Comment by ioanmo226** [comment]
Yes. It's working fine now

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [reply extension/js/common/api/email-provider/gmail/google-auth.ts:255]: Would be good to have a test for this, too. Since we use mocked tests, it should be straight forward. Could leave the test for another issue, but the checking needs to be fixed.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:127]: It seems you are probably creating `state` twice, once for request and once for verification?

It would be more natural to have a method `GoogleAuth.newAuthRequest(...) and then `state` would be one of the values in `AuthReq`. Then `getAuthRes` would have `expectedState: authRequest.state`.

It would better communicate the intetion for csrf protection, which is "I decide the value in the beginning and it must come back the same". Computing it twice obscures that intention.

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (tomholub) [inline extension/js/common/api/email-provider/gmail/google-auth.ts:196]: `Assert.urlParamRequire` throws error text that could be confusing for users, as it was meant for parsing URLs that we created ourselves. You'll notice this when reading the error text in `abortAndRenderErrOnUrlParamValMismatch`. It's not well documented.

But it's good to check it here. You could put these 4 lines in a try/catch, and if they throw error, return `        return { acctEmail, result: 'Error', error: `Could not parse URL returned from Google`, id_token: undefined };`

**Review by tomholub — COMMENTED** [review]
Looking good, see last few comments from my side :+1:

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (ioanmo226) [reply extension/js/common/api/email-provider/gmail/google-auth.ts:255]: UI test added

File: extension/js/common/api/email-provider/gmail/google-auth.ts
Comment (ioanmo226) [reply extension/js/common/api/email-provider/gmail/google-auth.ts]: Ok, Done.
I agree with you. but not sure exactly.

**Comment by ioanmo226** [comment]
By the way, @tomholub How are you gonna handle 2 other PRs(branches)?
Just close them and leave branches or what?

**Comment by tomholub** [comment]
yeah we'll just close them. once in a few years i clean up stale branches

File: extension/js/common/assert.ts
Comment (tomholub) [inline extension/js/common/assert.ts:17]: `AssertError` should inherit from `UnreportableError` so we don't end up having these reported to our error collection endpoint.

File: test/source/test.ts
Comment (tomholub) [inline test/source/test.ts:133]: That's what I meant. I don't want this reported, therefore this should be removed.

**Comment by tomholub** [comment]
Thanks for the added test. After my last round of comments I expect to merge this. Since this is important functionality, if @rrrooommmaaa has any comments after that, we should still file issues for them and resolve them in the same version before we do a release.

File: test/source/test.ts
Comment (ioanmo226) [reply test/source/test.ts:133]: Even though, `AssertError` inherits from `UnreportableError`, above code is necessary because of this line.
It returns result as error and so javascript error is occurred
```ts
if (receivedState !== expectedState) {
  return { acctEmail, result: 'Error', error: `Wrong oauth CSRF token. Please try again.`, id_token: undefined };
}
```

Comment (tomholub) [reply test/source/test.ts:133]: But it affects other code, as that error is thrown in many other places. We are not testing that, but there are other places using `Assert.` that will now newly be throwing reportable errors, which is not intended.

Comment (tomholub) [reply test/source/test.ts:133]: Anyway, please help me remove these two lines in test.ts. Rest looks good.

Comment (ioanmo226) [reply test/source/test.ts:133]: Yeah. already inherited `AssertError` from `UnreportableError`.
I think you misunderstood me.
Those above 2 lines are not needed because of AssertError not inherit from UnreportableError
It's needed because we check `crsftoken`(state) and return error if it doesn't match.

Comment (ioanmo226) [reply test/source/test.ts:133]: This line
```ts
return { acctEmail, result: 'Error', error: `Wrong oauth CSRF token. Please try again.`, id_token: undefined };
```

Comment (tomholub) [reply test/source/test.ts:133]: I see. I would not want that reported, either. It can sometimes happen maybe if user uses the back and forward button on the host page and then re-confirms the old oauth form, that kind of thing. Could this be selectively not reported?

Comment (ioanmo226) [reply test/source/test.ts:133]: Done

**Review by tomholub — APPROVED** [review]

**Comment by tomholub** [comment]
@ioanmo226 thank you :1st_place_medal: this was pending for years, but now it has become a real problem when Google decided to pull the plug on OOB.

**Comment by ioanmo226** [comment]
@tomholub Thank you for your help too :)

**Pull request merged**
